controllers/v1beta1/build_qoshandler.go

```diff
@@ -52,18 +51,42 @@ func (r *LagoonBuildReconciler) whichBuildNext(ctx context.Context, opLog logr.L
 	if err := r.List(ctx, runningBuilds, listOption); err != nil {
 		return fmt.Errorf("Unable to list builds in the cluster, there may be none or something went wrong: %v", err)
 	}
+	buildsToStart := r.BuildQoS.MaxBuilds - len(runningBuilds.Items)
 	if len(runningBuilds.Items) >= r.BuildQoS.MaxBuilds {
 		// if the maximum number of builds is hit, then drop out and try again next time
 		if r.EnableDebug {
 			opLog.Info(fmt.Sprintf("Currently %v running builds, no room for new builds to be started", len(runningBuilds.Items)))
 		}
+		go r.processQueue(ctx, opLog, buildsToStart, true)
 		return nil
 	}
-	buildsToStart := r.BuildQoS.MaxBuilds - len(runningBuilds.Items)
 	if buildsToStart > 0 {
 		opLog.Info(fmt.Sprintf("Currently %v running builds, room for %v builds to be started", len(runningBuilds.Items), buildsToStart))
 		// if there are any free slots to start a build, do that here
-		listOption = (&client.ListOptions{}).ApplyOptions([]client.ListOption{
+		go r.processQueue(ctx, opLog, buildsToStart, false)
+	}
+	return nil
+}
+
+var runningProcessQueue bool
+
+// this is a processor for any builds that are currently `queued` status. all normal build activity will still be performed
+// this just allows the controller to update any builds that are in the queue periodically
+// if this ran on every single event, it would flood the queue with messages, so it is restricted using `runningProcessQueue` global
+// to only run the process at any one time til it is complete
+func (r *LagoonBuildReconciler) processQueue(ctx context.Context, opLog logr.Logger, buildsToStart int, limitHit bool) error {
+	// this should only ever be able to run one instance of at a time within a single controller
+	// this is because this process is quite heavy when it goes to submit the queue messages to the api
+	// the downside of this is that there can be delays with the messages it sends to the actual
+	// status of the builds, but build complete/fail/cancel will always win out on the lagoon-core side
+	// so this isn't that much of an issue if there are some delays in the messages
+	opLog = opLog.WithName("QueueProcessor")
```

Comment:
as an aside, not part of the review, but I wonder if it wouldn't be cool to have the logger itself contain the logic for whether debug is enabled or not, so you're just calling a function

just a thought really

```
	enableDebug := true

	logger := func(l string) {
		if enableDebug {
			fmt.Println(l)
		}
	}

	logger("hello world")

	enableDebug = false

	logger("shouldn't show up")
```

Reply:
to be clear, this is absolutely not part of the review, just me thinking out loud :D

I know we're calling subfunctions of the logger class, that complicates it. It's likely just simpler to do the damned `if`

Reply:
I didn't want to complicate things with that, so the `if` is just simpler